runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java

```diff
@@ -172,48 +176,53 @@ public DirectPipelineResult run(Pipeline originalPipeline) {
     }
     pipeline.replaceAll(defaultTransformOverrides());
     MetricsEnvironment.setMetricsSupported(true);
-    DirectGraphVisitor graphVisitor = new DirectGraphVisitor();
-    pipeline.traverseTopologically(graphVisitor);
+    try {
+      DirectGraphVisitor graphVisitor = new DirectGraphVisitor();
+      pipeline.traverseTopologically(graphVisitor);
 
-    @SuppressWarnings("rawtypes")
-    KeyedPValueTrackingVisitor keyedPValueVisitor = KeyedPValueTrackingVisitor.create();
-    pipeline.traverseTopologically(keyedPValueVisitor);
+      @SuppressWarnings("rawtypes")
+      KeyedPValueTrackingVisitor keyedPValueVisitor = KeyedPValueTrackingVisitor.create();
+      pipeline.traverseTopologically(keyedPValueVisitor);
 
-    DisplayDataValidator.validatePipeline(pipeline);
-    DisplayDataValidator.validateOptions(getPipelineOptions());
+      DisplayDataValidator.validatePipeline(pipeline);
+      DisplayDataValidator.validateOptions(getPipelineOptions());
 
-    DirectGraph graph = graphVisitor.getGraph();
-    EvaluationContext context =
-        EvaluationContext.create(
-            getPipelineOptions(),
-            clockSupplier.get(),
-            Enforcement.bundleFactoryFor(enabledEnforcements, graph),
-            graph,
-            keyedPValueVisitor.getKeyedPValues());
+      DirectGraph graph = graphVisitor.getGraph();
+      ExecutorService metricsPool = Executors.newCachedThreadPool(
+          new ThreadFactoryBuilder()
+            .setThreadFactory(MoreExecutors.platformThreadFactory())
+            .setDaemon(false) // otherwise you say you want to leak, please don't!
+            .setNameFormat("direct-metrics-counter-committer")
+            .build());
+      EvaluationContext context = EvaluationContext.create(
+          getPipelineOptions(), clockSupplier.get(),
+          Enforcement.bundleFactoryFor(enabledEnforcements, graph),
+          graph, keyedPValueVisitor.getKeyedPValues(),
+          metricsPool);
 
-    TransformEvaluatorRegistry registry = TransformEvaluatorRegistry.defaultRegistry(context);
-    PipelineExecutor executor =
-        ExecutorServiceParallelExecutor.create(
-            options.getTargetParallelism(),
-            registry,
-            Enforcement.defaultModelEnforcements(enabledEnforcements),
-            context);
-    executor.start(graph, RootProviderRegistry.defaultRegistry(context));
+      TransformEvaluatorRegistry registry = TransformEvaluatorRegistry.defaultRegistry(context);
+      PipelineExecutor executor = ExecutorServiceParallelExecutor.create(
+          options.getTargetParallelism(), registry,
```

Comment:
nit: this is a bunch of no-op changes that just reformats things (to a worse format).